src/main/java/org/jvnet/solaris/libzfs/LibZFS.java

```diff
@@ -39,17 +39,122 @@
+        }
+        return "";
+    }
+
+    private String libzfs4j_get_extsetting(String libzfs4j_envvar_name) {
+        /* Ensure the caller has a string set via envvars or Java CLI.
+         * If the value is not set or is set to empty string, try the
+         * next option. If it ends up not set, return a "<NULL>". */
+        String libzfs4j_envvar_value;
+        libzfs4j_envvar_value = System.getenv(libzfs4j_envvar_name);
+        if (libzfs4j_envvar_value == null || libzfs4j_envvar_value.equals("")) {
+            libzfs4j_envvar_value = System.getProperty(libzfs4j_envvar_name);
+            if (libzfs4j_envvar_value == null || libzfs4j_envvar_value.equals("")) {
+                libzfs4j_envvar_value = "<NULL>";
+            }
+        }
+        return libzfs4j_envvar_value;
+    }
+
+    private void libzfs4j_add_feature(String libzfs4j_envvar_name, String libzfs4j_usevar_value, String libzfs4j_envvar_value) {
+        /* Add the key=value, reporting on it in log (and reporting the
+         * original envvvar/property value) for debugging. */
+        /* Note: we do not slow down to replace-if-present logic, since
+         * by current implementation and intention, this is only used from
+         * libzfs4j_init_features() below and that one only runs once.
+         * This decision can be amended later, if ever needed. */
+        Entry<String, String> myEntry;
+
+        System.out.println("[LIBZFS4J-DEBUG]: Setting '" +
+            libzfs4j_envvar_name + "' = '" + libzfs4j_usevar_value +
+            "' (envvar value was '" + libzfs4j_envvar_value + "')");
+        myEntry = new SimpleEntry<String, String>(libzfs4j_envvar_name, libzfs4j_usevar_value);
+        libzfs_features.add(myEntry);
+    }
+
+    void libzfs4j_init_features() {
+        String libzfs4j_envvar_value;
+        String libzfs4j_envvar_name;
+        String libzfs4j_usevar_value;
+        String libzfs4j_default_abi; /* Cache the default while we make decisions */
+
+        if (libzfs_features.size() > 0)
+            return; /* already inited */
+
+        libzfs4j_envvar_name = "LIBZFS4J_ABI";
+        libzfs4j_envvar_value = libzfs4j_get_extsetting(libzfs4j_envvar_name);
+        if (libzfs4j_envvar_value.equals("legacy") || libzfs4j_envvar_value.equals("openzfs")) {
+            /* Currently we recognize two values; later it may be more like openzfs-YYYY */
+            libzfs4j_default_abi = libzfs4j_envvar_value;
+        } else {
+            /* Detect presence of e.g. feature flags routines == openzfs */
+            libzfs4j_default_abi = "legacy";
+        }
+        libzfs4j_usevar_value = libzfs4j_default_abi;
+        libzfs4j_add_feature(libzfs4j_envvar_name, libzfs4j_usevar_value, libzfs4j_envvar_value);
+
+        libzfs4j_envvar_name = "LIBZFS4J_ABI_zfs_iter_snapshots";
+        libzfs4j_envvar_value = libzfs4j_get_extsetting(libzfs4j_envvar_name);
+        if (libzfs4j_envvar_value.equals("legacy") || libzfs4j_envvar_value.equals("3")) {
```

Comment:
Maybe you want to use case-insensitive comparisons. just in case somebody passes "Legacy"